File: attempt1/lib/src/neuron.dart
part of nnn;

abstract class Neuron {
  List<Neuron> destinations = [];
  List<Neuron> inputs = [];
  bool _willFire = false;
  bool _firing = false;
  
  bool get firing => _firing;
  
  int get messageCount {
    int count = 0;
    for (var input in inputs) {
      if (input._firing) {
        ++count;
      }
    }
    return count;
  }
  
  bool nextCycle();
}

class InputNeuron extends Neuron {
  bool _fireNext = false;
  
  void trigger() {
    _fireNext = true;
  }
  
  bool nextCycle() {
    bool result = _fireNext;
    _fireNext = false;
    return result;
  }
}

class OutputNeuron extends Neuron {
  bool _triggered = false;
  
  bool get triggered => _triggered;
  
  bool nextCycle() {
    _triggered = (messageCount != 0);
    return false;
  }
}

class DelayNeuron extends Neuron {
  bool nextCycle() {
    return messageCount != 0;
  }
}

class AndNeuron extends Neuron {
  bool nextCycle() {
    return messageCount == 2;
  }
}

class OrNeuron extends Neuron {
  bool nextCycle() {
    return messageCount > 0;
  }
}

class XorNeuron extends Neuron {
  bool nextCycle() {
    return messageCount == 1;
  }
}

File: attempt1/go-attempt/src/neuron.go
package nnn

const (
	NEURON_XOR = iota
	NEURON_AND = iota
	NEURON_OR  = iota
)

type Neuron struct {
	inputs   []*Link
	outputs  []*Link
	firing   bool
	willFire bool
	function int
}

func (self *Neuron) InputCount() int {
	count := 0
	for input := range self.inputs {
		if input.firing {
			count++
		}
	}
	return count
}

func (self *Neuron) NextCycle() bool {
	switch self.function {
	case NEURON_XOR:
		return self.InputCount()%2 != 0
	case NEURON_AND:
		for input := range self.inputs {
			if !input.firing {
				return false
			}
		}
		return true
	case NEURON_OR:
		for input := range self.inputs {
			if input.firing {
				return true
			}
		}
		return false
	}
}
